Building with a custom Gradle build type fails if plugins are present.

Cause of the errors behind https://github.com/flutter/flutter/pull/71660

Adding the following to app/build.gradle:

```gradle
android {
    buildTypes {
        local {
            initWith debug
        }
    }
}
```

To any flutter project with a plugin will fail to build.

/cc @blasten who hopefully knows how to fix this off the top of his head (:D)

Sample error output:

```
./gradlew assembleLocal

FAILURE: Build failed with an exception.

* What went wrong:
Could not determine the dependencies of task ':app:compileLocalJavaWithJavac'.
> Could not resolve all task dependencies for configuration ':app:localCompileClasspath'.
   > Could not resolve project :integration_test.
     Required by:
         project :app
      > No matching variant of project :integration_test was found. The consumer was configured to find an API of a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'local', attribute 'org.jetbrains.kotlin.platform.type' with value 'androidJvm' but:
          - Variant 'debugApiElements' capability com.example.integration_test:integration_test:1.0-SNAPSHOT declares an API of a component:
              - Incompatible because this component declares a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'debug' and the consumer needed a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'local'
              - Other compatible attribute:
                  - Doesn't say anything about org.jetbrains.kotlin.platform.type (required 'androidJvm')
          - Variant 'debugRuntimeElements' capability com.example.integration_test:integration_test:1.0-SNAPSHOT declares a runtime of a component:
              - Incompatible because this component declares a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'debug' and the consumer needed a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'local'
              - Other compatible attribute:
                  - Doesn't say anything about org.jetbrains.kotlin.platform.type (required 'androidJvm')
          - Variant 'profileApiElements' capability com.example.integration_test:integration_test:1.0-SNAPSHOT declares an API of a component:
              - Incompatible because this component declares a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'profile' and the consumer needed a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'local'
              - Other compatible attribute:
                  - Doesn't say anything about org.jetbrains.kotlin.platform.type (required 'androidJvm')
          - Variant 'profileRuntimeElements' capability com.example.integration_test:integration_test:1.0-SNAPSHOT declares a runtime of a component:
              - Incompatible because this component declares a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'profile' and the consumer needed a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'local'
              - Other compatible attribute:
                  - Doesn't say anything about org.jetbrains.kotlin.platform.type (required 'androidJvm')
          - Variant 'releaseApiElements' capability com.example.integration_test:integration_test:1.0-SNAPSHOT declares an API of a component:
              - Incompatible because this component declares a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'release' and the consumer needed a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'local'
              - Other compatible attribute:
                  - Doesn't say anything about org.jetbrains.kotlin.platform.type (required 'androidJvm')
          - Variant 'releaseRuntimeElements' capability com.example.integration_test:integration_test:1.0-SNAPSHOT declares a runtime of a component:
              - Incompatible because this component declares a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'release' and the consumer needed a component, as well as attribute 'com.android.build.api.attributes.BuildTypeAttr' with value 'local'
              - Other compatible attribute:
                  - Doesn't say anything about org.jetbrains.kotlin.platform.type (required 'androidJvm')

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.7/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 1s
```

--- a/dev/devicelab/bin/tasks/gradle_plugin_light_apk_test.dart
+++ b/dev/devicelab/bin/tasks/gradle_plugin_light_apk_test.dart
@@ -8,6 +8,7 @@ import 'package:flutter_devicelab/framework/apk_utils.dart';
 import 'package:flutter_devicelab/framework/framework.dart';
 import 'package:flutter_devicelab/framework/task_result.dart';
 import 'package:flutter_devicelab/framework/utils.dart';
+import 'package:path/path.dart' as path;
 
 Future<void> main() async {
   await task(() async {
@@ -204,6 +205,35 @@ Future<void> main() async {
       await runProjectTest((FlutterProject project) async {
         section('gradlew assembleLocal (custom debug build)');
         await project.addCustomBuildType('local', initWith: 'debug');
+        await project.runGradleTask('assembleLocal');
+      });
+
+      await runProjectTest((FlutterProject project) async {
+        section('gradlew assembleLocal with plugin (custom debug build)');
+
+        final Directory tempDir = Directory.systemTemp.createTempSync('flutter_plugin.');
+        final Directory pluginDir = Directory(path.join(tempDir.path, 'plugin_under_test'));
+
+        section('Create plugin');
+        await inDirectory(tempDir, () async {
+          await flutter(
+            'create',
+            options: <String>[
+              '--org',
+              'io.flutter.devicelab.plugin',
+              '--template=plugin',
+              '--platforms=android,ios',
+              pluginDir.path,
+            ],
+          );
+        });
+
+        section('Configure');
+        project.addPlugin('plugin_under_test', value: '$platformLineSep    path: ${pluginDir.path}');
+        await project.addCustomBuildType('local', initWith: 'debug');
+        await project.getPackages();
+
+        section('Build APK');
         await project.runGradleTask('assembleLocal');
       });
 

--- a/dev/devicelab/lib/framework/apk_utils.dart
+++ b/dev/devicelab/lib/framework/apk_utils.dart
@@ -282,7 +282,7 @@ subprojects {
 
   /// Adds a plugin to the pubspec.
   /// In pubpsec, each dependency is expressed as key, value pair joined by a colon `:`.
-  /// such as `plugin_a`:`^0.0.1` or `path`:`\npath: /some/path`.
+  /// such as `plugin_a`:`^0.0.1` or `plugin_a`:`\npath: /some/path`.
   Future<void> addPlugin(String plugin, { String value = '' }) async {
     final File pubspec = File(path.join(rootPath, 'pubspec.yaml'));
     String content = await pubspec.readAsString();

--- a/dev/devicelab/lib/framework/apk_utils.dart
+++ b/dev/devicelab/lib/framework/apk_utils.dart
@@ -280,12 +280,15 @@ subprojects {
     ''');
   }
 
-  Future<void> addPlugin(String plugin) async {
+  /// Adds a plugin to the pubspec.
+  /// In pubpsec, each dependency is expressed as key, value pair joined by a colon `:`.
+  /// such as `plugin_a`:`^0.0.1` or `path`:`\npath: /some/path`.
+  Future<void> addPlugin(String plugin, { String value = '' }) async {
     final File pubspec = File(path.join(rootPath, 'pubspec.yaml'));
     String content = await pubspec.readAsString();
     content = content.replaceFirst(
       '${platformLineSep}dependencies:$platformLineSep',
-      '${platformLineSep}dependencies:$platformLineSep  $plugin:$platformLineSep',
+      '${platformLineSep}dependencies:$platformLineSep  $plugin:$value$platformLineSep',
     );
     await pubspec.writeAsString(content, flush: true);
   }

--- a/dev/devicelab/lib/framework/apk_utils.dart
+++ b/dev/devicelab/lib/framework/apk_utils.dart
@@ -281,7 +281,7 @@ subprojects {
   }
 
   /// Adds a plugin to the pubspec.
-  /// In pubpsec, each dependency is expressed as key, value pair joined by a colon `:`.
+  /// In pubspec, each dependency is expressed as key, value pair joined by a colon `:`.
   /// such as `plugin_a`:`^0.0.1` or `plugin_a`:`\npath: /some/path`.
   Future<void> addPlugin(String plugin, { String value = '' }) async {
     final File pubspec = File(path.join(rootPath, 'pubspec.yaml'));

--- a/packages/flutter_tools/gradle/flutter.gradle
+++ b/packages/flutter_tools/gradle/flutter.gradle
@@ -149,19 +149,16 @@ class FlutterPlugin implements Plugin<Project> {
         String flutterExecutableName = Os.isFamily(Os.FAMILY_WINDOWS) ? "flutter.bat" : "flutter"
         flutterExecutable = Paths.get(flutterRoot.absolutePath, "bin", flutterExecutableName).toFile();
 
-        // Add custom build types.
+        String flutterProguardRules = Paths.get(flutterRoot.absolutePath, "packages", "flutter_tools",
+                "gradle", "flutter_proguard_rules.pro")
         project.android.buildTypes {
+            // Add profile build type.
             profile {
                 initWith debug
                 if (it.hasProperty("matchingFallbacks")) {
                     matchingFallbacks = ["debug", "release"]
                 }
             }
-        }
-
-        String flutterProguardRules = Paths.get(flutterRoot.absolutePath, "packages", "flutter_tools",
-                "gradle", "flutter_proguard_rules.pro")
-        project.android.buildTypes {
             release {
                 // Enables code shrinking, obfuscation, and optimization for only
                 // your project's release build type.
@@ -304,8 +301,10 @@ class FlutterPlugin implements Plugin<Project> {
             return
         }
         // Add plugin dependency to the app project.
+        // Use the default configuration, so the app project can define custom configurations
+        // (such as those created after adding a flavor or a custom build type).
         project.dependencies {
-            implementation pluginProject
+            implementation project(path: ":$pluginName", configuration: 'default')
         }
         Closure addEmbeddingCompileOnlyDependency = { buildType ->
             String flutterBuildMode = buildModeFor(buildType)

--- a/packages/flutter_tools/gradle/flutter.gradle
+++ b/packages/flutter_tools/gradle/flutter.gradle
@@ -315,18 +315,16 @@ class FlutterPlugin implements Plugin<Project> {
             // Copy build types from the app to the plugin.
             // This allows to build apps with plugins and custom build types or flavors.
             pluginProject.android.buildTypes {
-              "${buildType.name}" {}
+                "${buildType.name}" {}
             }
-            addApiDependencies(
-                pluginProject,
-                buildType.name,
-                "io.flutter:flutter_embedding_$flutterBuildMode:$engineVersion"
-            )
+            pluginProject.dependencies.add(
+                "${buildType.name}CompileOnly",
+                "io.flutter:flutter_embedding_$flutterBuildMode:$engineVersion")
         }
         // Wait until the Android plugin loaded.
         pluginProject.afterEvaluate {
-          project.android.buildTypes.each addEmbeddingCompileOnlyDependency
-          project.android.buildTypes.whenObjectAdded addEmbeddingCompileOnlyDependency
+            project.android.buildTypes.each addEmbeddingCompileOnlyDependency
+            project.android.buildTypes.whenObjectAdded addEmbeddingCompileOnlyDependency
         }
     }
 

--- a/packages/flutter_tools/gradle/flutter.gradle
+++ b/packages/flutter_tools/gradle/flutter.gradle
@@ -301,10 +301,8 @@ class FlutterPlugin implements Plugin<Project> {
             return
         }
         // Add plugin dependency to the app project.
-        // Use the default configuration, so the app project can define custom configurations
-        // (such as those created after adding a flavor or a custom build type).
         project.dependencies {
-            implementation project(path: ":$pluginName", configuration: 'default')
+            implementation pluginProject
         }
         Closure addEmbeddingCompileOnlyDependency = { buildType ->
             String flutterBuildMode = buildModeFor(buildType)
@@ -314,20 +312,21 @@ class FlutterPlugin implements Plugin<Project> {
             if (!supportsBuildMode(flutterBuildMode)) {
                 return
             }
+            // Copy build types from the app to the plugin.
+            // This allows to build apps with plugins and custom build types or flavors.
+            pluginProject.android.buildTypes {
+              "${buildType.name}" {}
+            }
             addApiDependencies(
                 pluginProject,
                 buildType.name,
                 "io.flutter:flutter_embedding_$flutterBuildMode:$engineVersion"
             )
         }
+        // Wait until the Android plugin loaded.
         pluginProject.afterEvaluate {
-          pluginProject.android.buildTypes {
-              profile {
-                  initWith debug
-              }
-          }
-          pluginProject.android.buildTypes.each addEmbeddingCompileOnlyDependency
-          pluginProject.android.buildTypes.whenObjectAdded addEmbeddingCompileOnlyDependency
+          project.android.buildTypes.each addEmbeddingCompileOnlyDependency
+          project.android.buildTypes.whenObjectAdded addEmbeddingCompileOnlyDependency
         }
     }
 

